Add an option to automatically sync the color mode with the operating system preference

akabiru opened the pull request:
# Ticket
<!-- Provide the link to respective work package -->

<!-- Contributors: Please check our PR guide: https://www.openproject.org/docs/development/code-review-guidelines/#preparing-your-pull-request before opening a PR. -->

<!-- Reviewers: Please check our Review guide: https://www.openproject.org/docs/development/code-review-guidelines/#reviewing -->

https://community.openproject.org/wp/66397

# What are you trying to accomplish?
<!-- Provide a description of the changes. -->

Adds an option to automatically sync the color mode with the operating system preference including contrast (only light high contrast supported as of this PR)

**Reviewer note:** _Supporting "dark high contrast" and separating "contrast settings" will be in separate PRs_

## Screenshots
<!-- Provide before/after screenshots, videos, or graphs for any visual changes; otherwise, remove this section -->

[auto-theme-switcher.webm](https://github.com/user-attachments/assets/b479dc61-75aa-4456-b832-147b0c8a0819)

# What approach did you choose and why?
<!-- This section is a place for you to describe your thought process in making these changes.
     List any tradeoffs you made to take on or pay down tech debt.
     Describe any alternative approaches you considered and why you discarded them. -->

Leverage the `matchMedia` browser API to detect the user preference and set the corresponding theme: (1) as an inline script in `head` to run before the body renders. This prevents flickering as the theme as updated. (2) An `auto-theme-switcher` stimulus controller listening to OS theme & contrast changes via [useMatchMedia](https://github.com/stimulus-use/stimulus-use/blob/main/docs/use-match-media.md) hooks

# Merge checklist

- [x] Added/updated tests
- [ ] Added/updated documentation in Lookbook (patterns, previews, etc)
- [x] Tested major browsers (Chrome, Firefox, Edge, ...)

akabiru committed:
Add Auto Sync Color Theme with System (OS) user preference

akabiru committed:
Fix theme flickering for sync_with_system preference

When using sync_with_system theme preference, there was a visible flicker
as the page defaulted to light theme before the Stimulus controller
could detect and apply the correct system theme.

This fix adds inline JavaScript in the layout head that detects the
system theme preference and applies it immediately before the body
renders, eliminating the flickering issue.

akabiru committed:
Rename option to `Auto (system default)`

akabiru committed:
Rename sync_with_os -> "Automatic (match OS colour mode)"

akabiru committed:
Add feature tests for automatic theme switcher

akabiru committed:
Automatically detect user contrast setting

akabiru committed:
Remove flaky high contrast feature test

This relies on the OS contrast setting which is not straightforward to replicate in CI

akabiru commented:
@HDinger @bsatarnejad  I tried to add a feature test for contrast detection in https://github.com/opf/openproject/pull/19746/commits/15475aec8ec62568cb29062b7e3d60dbc2c1fe46, but it fails on CI as it's dependent on the OS contrast settings.

Browser drivers don't seem to have an obvious override that is independent of the underlying OS preference. Apparently, chrome driver does even not support this as well. Left it out for now.

HDinger commented on frontend/src/stimulus/controllers/auto-theme-switcher.controller.ts:
```diff
@@ -0,0 +1,83 @@
+  connect() {
+    if (this.modeValue !== 'sync_with_os') return;
+
+    useMatchMedia(this, {
```
TIL 💡 Nice 👍

HDinger commented on app/views/layouts/base.html.erb:
```diff
@@ -40,6 +40,32 @@ See COPYRIGHT and LICENSE files for more details.
+    <%= nonced_javascript_tag do %>
+      // Apply system theme before body renders to prevent flickering
+      // This duplicates logic from AutoThemeSwitcher controller for immediate execution
+      (function() {
```
I'm really not a fan of that duplication. Out of the top of my head, I don't know how to avoid it. We can probably either extract it into a helper which we use here and in the controller or we find a way to call the stimulus controller directly.
That is already another nonced_javascript_tag at the bottom bottom of the file. Maybe we can match them together to only have one.

akabiru replied:
Thanks for the review @HDinger , I wasn't sure how to approach this issue initially but I think I've got a sensible solution now.

Please see: https://github.com/opf/openproject/pull/19746/commits/ede66462a59a01d2cb8c0530b17fb3833b316544, https://github.com/opf/openproject/pull/19746/commits/63c32e562c6b6c7afecbfaa721545785dfa324b6

> Encapsulate theme utilities into a class, with instances on the `window.OpenProject` and `AutoThemeSwitcher` interfaces.
> 
> AFAICT, it's not feasible to group the theme application in `base.html.erb` into the deffered `nonced_javascript_tag` as the deffered script would run too late for theme application. We also don't always need load this script `if User.current.pref.sync_with_system_theme?` is `false`

HDinger reviewed (commented):
Very nice, works like charm 👏 🙇 

I only have one comment which is not a blocker however. I think it would be better to get rid of the duplication but if you do not find a way, I'm also fine with it

akabiru committed:
DRY up theme switching implementation

Address code duplication concern raised in https://github.com/opf/openproject/pull/19746#discussion_r2256365160.

Encapsulate theme utilities into a class, with instances on the `window.OpenProject` and `AutoThemeSwitcher` interfaces.

AFAICT, it's not feasible to group the theme application in `base.html.erb` into the deffered `nonced_javascript_tag` as the
deffered script would run too late for theme application.

akabiru committed:
Reuse `window.OpenProject.theme` instance

HDinger reviewed (approved):
Nice 👏
